### PR TITLE
Cleanup in preparation for typefeedback

### PR DESCRIPTION
**Pull request by o-**

### CODE REVIEW

**Comment by o-** [comment]
A Bunch of cleanups, in preparation for collecting typefeedback.

I wanted to simplify the Code struct. For that I removed the CallSite struct part. This means that I put the main parts of the CallSite struct into the BC stream, and the non-essential parts I threw away for now.

I think this makes a lot more sense, but this also means, that i had to add 2 BC and 3 BCs became variable length. On the plus side, the BC objects from `ir/BC_inc.h` now really capture the whole BC, with all arguments, and it is no longer necessary to pass around CallSite pointers with it.

Also, I removed the CodeEditor, since that one was quite intertwined with the CallSite structs.

The idea is to later add a new (and better designed) part in the Code object, where we can store additional metadata for any BC. But the idea is, that the metadata is not necessary to interpret the behavior of a BC!

File: rir/src/runtime/Code.h [hunk 1]
```diff
@@ -202,7 +202,6 @@ struct CallSite {
 
     uint32_t hasNames : 1;
     uint32_t hasTarget : 1;
-    uint32_t hasImmediateArgs : 1;
```
Comment (JanJecmen) [inline rir/src/runtime/Code.h]: `free` should be 30 now, but not sure if it even matters

File: rir/src/interpreter/interp.cpp [hunk 2]
```diff
@@ -1427,9 +1425,12 @@ SEXP evalRirCode(Code* c, Context* ctx, SEXP* env,
...
             Immediate id = readImmediate();
             advanceImmediate();
-            advanceImmediate(); // nargs, TODO: remove
-            CallContext call(c, id, ostack_top(ctx), getenv(), ctx);
+            size_t n = readImmediate();
+            advanceImmediate();
+            CallContext call(c, id, ostack_top(ctx), (Immediate*)pc, getenv(),
+                             ctx);
             res = doCall(call, ctx);
+            advanceImmediateN(n);
```
Comment (JanJecmen) [inline rir/src/interpreter/interp.cpp]: Why is this after the call?

File: rir/src/compiler/translations/pir_2_rir.cpp [hunk 3]
```diff
@@ -923,29 +923,28 @@ size_t Pir2Rir::compileCode(Context& ctx, Code* code) {
 
             case Tag::Call: {
                 auto call = Call::Cast(instr);
-                cs.insertCall(Opcode::call_, call->nCallArgs(), {},
-                              Pool::get(call->srcIdx));
+                cs.insertCall(call->nCallArgs(), Pool::get(call->srcIdx));
```
Comment (JanJecmen) [inline rir/src/compiler/translations/pir_2_rir.cpp]: So, a bit unrelated, but any reason why the src indices for calls are in the constant pool and not the source pool?
(and also why the source pool does not manage duplicities?)

File: rir/src/interpreter/interp.cpp [hunk 4]
```diff
@@ -73,6 +90,11 @@ struct CallContext {
         return ostack_at_cell(stackArgs + i);
     }
 
+    SEXP name(unsigned i, Context* ctx) const {
```
Comment (JanJecmen) [inline rir/src/interpreter/interp.cpp:91]: Another unrelated question: why do we pass around the Context pointers everywhere, when the context is only ever the globalContext?

File: rir/src/R/Printing.h [hunk 5]
```diff
@@ -0,0 +1,17 @@
+#ifndef RIR_R_PRINT
+#define RIR_R_PRINT
+
+#include "R/r.h"
+#include "utils/capture_out.h"
+
+namespace rir {
+
+static std::string dumpSexp(SEXP src, size_t length = 50) {
```
Comment (JanJecmen) [inline rir/src/R/Printing.h:9]: Maybe it would be useful to also include the type of the sexp in the dump? But not sure..

**Review by JanJecmen — APPROVED** [review]
I like the simplifications!

File: rir/src/runtime/Code.h (hunk 1 above)
Comment (o-) [reply rir/src/runtime/Code.h]: uhm, that was not the final version :) its gone now

File: rir/src/compiler/translations/pir_2_rir.cpp (hunk 3 above)
Comment (o-) [reply rir/src/compiler/translations/pir_2_rir.cpp]: ehm, no, not really.

File: rir/src/interpreter/interp.cpp (hunk 4 above)
Comment (o-) [reply rir/src/interpreter/interp.cpp:91]: yeah, there was once the idea to have it ready for multple contexts, i was not involved.

File: rir/src/R/Printing.h (hunk 5 above)
Comment (o-) [reply rir/src/R/Printing.h:9]: I mostly use it for the disassembly, and there i mainly care about it not being cluttered. for other cases maybe...

**Pull request merged**